src/Domain/CoordinateMaps/FocallyLiftedMapHelpers.cpp

```diff
@@ -0,0 +1,215 @@
+      x1 = 1.0;
+    }
+    if (pick_root_greater_than_one) {
+      // For the inverse map, we want the a scale_factor s such that
+      // s >= 1. Note that gsl_poly_solve_quadratic returns x0 < x1.
+      // have three cases:
+      //  a) x0 < x1 < 1         ->   error
+      //  b) x0 < 1 <= x1        ->   Choose x1
+      //  c) 1 <= x0 < x1        ->   choose based on pick_larger_root
+      if (x0 >= 1.0 and not pick_larger_root) {
+        return x0;
+      } else if (x1 >= 1.0) {
+        return x1;
+      } else {
+        return boost::none;
+      }
+    } else {
+      // For the inverse map, we want a scale_factor s such that 0 < s <= 1.
+      // Note that gsl_poly_solve_quadratic returns x0 < x1.
+      // So we have six cases:
+      //  a) x0 < x1 <= 0        ->   error
+      //  b) x0 <= 0 < x1 <= 1   ->   Choose x1
+      //  c) x0 <= 0 and x1 > 1  ->   error
+      //  d) 0 < x0 < x1 <= 1      ->   Choose according to pick_larger_root
+      //  e) 0 < x0 <= 1 < x1      ->   Choose x0
+      //  f) 1 < x0 < x1           ->   error
+      if (x0 <= 0.0) {
+        if (x1 > 0.0 and x1 <= 1.0) {
+          return x1;  // b)
+        } else {
+          return boost::none;  // a) and c)
+        }
+      } else if (x0 <= 1.0) {
+        if (x1 > 1.0) {
+          return x0;  // e)
+        } else {
+          return pick_larger_root ? x1 : x0;  // d)
+        }
+      } else {
+        return boost::none;  // f)
+      }
+    }
+  } else if (num_real_roots == 1) {
+    // We solved for scale_factor-1 above, so add 1 to get scale_factor.
+    x0 += 1.0;
+    if (equal_within_roundoff(x0, 1.0)) {
+      x0 = 1.0;
+    }
+    if (pick_root_greater_than_one) {
+      if (x0 < 1.0) {
+        return boost::none;
+      } else {
+        return x0;
+      }
+    } else {
+      if (x0 <= 0.0 or x0 > 1.0) {
+        return boost::none;
+      } else {
+        return x0;
+      }
+    }
+  } else {
+    return boost::none;
+  }
+}
+
+template <typename T>
+std::array<tt::remove_cvref_wrap_t<T>, 3> d_scale_factor_d_src_point(
+    const std::array<T, 3>& intersection_point,
+    const std::array<double, 3>& proj_center,
+    const std::array<double, 3>& sphere_center, const T& lambda) noexcept {
+  using return_type = tt::remove_cvref_wrap_t<T>;
+  const return_type lambda_over_denominator =
+      square(lambda) / (square(intersection_point[0] - proj_center[0]) +
+                        square(intersection_point[1] - proj_center[1]) +
+                        square(intersection_point[2] - proj_center[2]) +
+                        ((intersection_point[0] - proj_center[0]) *
+                             (proj_center[0] - sphere_center[0]) +
+                         (intersection_point[1] - proj_center[1]) *
+                             (proj_center[1] - sphere_center[1]) +
+                         (intersection_point[2] - proj_center[2]) *
+                             (proj_center[2] - sphere_center[2])));
```

Comment:
Are you in trouble if projection point == sphere center or intersection point == projection point? Could that ever happen?